### PR TITLE
Cmr 447

### PR DESCRIPTION
**Pull request by dseelig**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 /target/
+src/main/java/de/dnb
```
Comment (claussni) [inline .gitignore]: Why is a source folder in .gitignore?

File: pom.xml [hunk 2]
```diff
@@ -1,29 +1,20 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (claussni) [inline pom.xml]: Unnecessary header formatting.

File: pom.xml [hunk 3]
```diff
@@ -103,12 +96,14 @@
                 <plugin>
                     <groupId>org.apache.maven.plugins</groupId>
                     <artifactId>maven-compiler-plugin</artifactId>
-                    <version>3.3</version>
+                    <version>3.7.0</version>
                     <configuration>
                         <source>1.7</source>
                         <target>1.7</target>
+                        <compilerArgument>-proc:none</compilerArgument>
```
Comment (claussni) [inline pom.xml]: What is the purpose of this compiler argument?

File: src/main/java/de/qucosa/dissemination/epicur/EpicurDissMapper.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    private String transferUrlPattern;
+
+    private String frontpageUrlPattern;
+
+    private boolean transferUrlPidencode;
+
+    private static final String XEPICUR_SCHEMA_LOCATION = "urn:nbn:de:1111-2004033116 http://www.persistent-identifier.de/xepicur/version1.0/xepicur.xsd";
+
+    @SuppressWarnings("serial")
```
Comment (claussni) [inline src/main/java/de/qucosa/dissemination/epicur/EpicurDissMapper.java]: This whole initializer is redundant. 
Btw. what is @SuppressWarnings("serial")?

File: src/main/java/de/qucosa/dissemination/epicur/EpicurDissMapper.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        {
+            put("ubc", "monarch");
+        }
+        {
+            put("ubl", "ul");
+        }
+    };
+
+    public EpicurDissMapper(String transferUrlPattern, String frontpageUrlPattern, String agentNameSubstitutions, boolean transferUrlPidencode)
+            throws JDOMException, IOException {
```
Comment (claussni) [inline src/main/java/de/qucosa/dissemination/epicur/EpicurDissMapper.java]: IntelliJ code analyzer shows that this two Exception are never thrown.

File: src/main/java/de/qucosa/dissemination/epicur/EpicurDissMapper.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            .frontpageUrlPattern(frontpageUrlPattern)
+            .mets(this.metsDoc)
+            .transferUrlPattern(transferUrlPattern)
+            .updateStatus(UpdateStatus.urn_new);
+        
+        DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+        dbf.setNamespaceAware(true);
+        DocumentBuilder db = dbf.newDocumentBuilder();
+
+        Document epicurDoc = null;
```
Comment (claussni) [inline src/main/java/de/qucosa/dissemination/epicur/EpicurDissMapper.java]: Redundant initialization...

File: src/test/java/de/qucosa/dissemination/epicur/mapper/EpicurMapperTests.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        
+        Document epicurRes = mapper.transformEpicurDiss(mets);
+        
+        OutputFormat outputFormat = new OutputFormat(epicurRes);
+        outputFormat.setOmitXMLDeclaration(true);
+        StringWriter stringWriter = new StringWriter();
+        XMLSerializer serialize = new XMLSerializer(stringWriter, outputFormat);
+        serialize.serialize(epicurRes);
+        
+        System.out.println(stringWriter.toString());
```
Comment (claussni) [inline src/test/java/de/qucosa/dissemination/epicur/mapper/EpicurMapperTests.java]: Test should assert things, not print out to console. That's not a test. That's a 'try' ;-)

File: src/test/resources/hochschulschrift_TEST.xml [hunk 8]
```diff
@@ -0,0 +1,90 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (claussni) [inline src/test/resources/hochschulschrift_TEST.xml]: (!) This file cannot stay as part of the code since it contains protected personal data.
If you need a test file, construct one by removing production information and only retain the document structure.

File: src/test/java/de/qucosa/dissemination/epicur/mapper/EpicurMapperTests.java (hunk 7 above)
Comment (dseelig) [reply src/test/java/de/qucosa/dissemination/epicur/mapper/EpicurMapperTests.java]: Remove print out from this method.
Add assert of epicur element node in parsed epicur document.

File: src/test/java/de/qucosa/dissemination/epicur/mapper/EpicurMapperTest.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        dbf.setNamespaceAware(true);
+        DocumentBuilder db = dbf.newDocumentBuilder();
+        Document mets = db.parse(getClass().getResourceAsStream("/hochschulschrift_TEST.xml"));
+        EpicurDissMapper mapper = new EpicurDissMapper("http://test.##AGENT##.qucosa.de/id/##PID##", "", "", true);
+        Document epicurRes = mapper.transformEpicurDiss(mets);
+        
+        Assert.assertEquals("epicur", epicurRes.getDocumentElement().getTagName());
+    }
+
+    private XMLSerializer serializeEpicur(Document document) {
```
Comment (claussni) [inline src/test/java/de/qucosa/dissemination/epicur/mapper/EpicurMapperTest.java]: Unused method?

File: src/test/resources/hochschulschrift_TEST.xml [hunk 10]
```diff
@@ -0,0 +1,89 @@
+<!--Generated by METS Dissemination (2.0.0)--><mets:mets OBJID="qucosa:12164" xsi:schemaLocation="http://www.loc.gov/METS/ http://www.loc.gov/standards/mets/mets.xsd" xmlns:mets="http://www.loc.gov/METS/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"><mets:metsHdr RECORDSTATUS="ACTIVE" CREATEDATE="2017-06-01T10:59:21.052+00:00" LASTMODDATE="2017-06-01T10:59:29.128+00:00"><mets:agent ROLE="EDITOR" TYPE="ORGANIZATION"><mets:name>ubl</mets:name></mets:agent></mets:metsHdr><mets:dmdSec ID="DMD_000" STATUS="ACTIVE"><mets:mdWrap MDTYPE="MODS"><mets:xmlData><v3:mods xmlns:v3="http://www.loc.gov/mods/v3">
```
Comment (claussni) [inline src/test/resources/hochschulschrift_TEST.xml]: Is this production data?  Anonymize the file! No real email address or birth dates are allowed.

File: src/test/resources/qucosa-48666.xml [hunk 11]
```diff
@@ -0,0 +1,159 @@
+<mets OBJID="qucosa:48666" schemaLocation="http://www.loc.gov/METS/ http://www.loc.gov/standards/mets/mets.xsd">
```
Comment (claussni) [inline src/test/resources/qucosa-48666.xml]: Anonymize!

File: pom.xml [hunk 12]
```diff
@@ -1,29 +1,19 @@
...
-  ~ Unless required by applicable law or agreed to in writing, software
-  ~ distributed under the License is distributed on an "AS IS" BASIS,
-  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-  ~ See the License for the specific language governing permissions and
-  ~ limitations under the License.
-  -->
-
-<project xmlns="http://maven.apache.org/POM/4.0.0"
-         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+<!-- ~ Copyright 2016 Saxon State and University Library Dresden (SLUB) ~
```
Comment (claussni) [inline pom.xml]: Unnecessary formatting change.

File: pom.xml [hunk 13]
```diff
@@ -94,6 +62,30 @@
...
+
+        <dependency>
+            <groupId>com.sun.xml.bind</groupId>
+            <artifactId>jaxb-core</artifactId>
+            <version>2.3.0</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.eclipse.birt.runtime.3_7_1</groupId>
+            <artifactId>org.apache.xml.serializer</artifactId>
```
Comment (claussni) [inline pom.xml]: Do we really need this XML serializer from the BIRT-Project? I'd rather go with JDKs Transformer as long as it is sufficient.

File: pom.xml [hunk 14]
```diff
@@ -103,12 +95,13 @@
                 <plugin>
                     <groupId>org.apache.maven.plugins</groupId>
                     <artifactId>maven-compiler-plugin</artifactId>
-                    <version>3.3</version>
+                    <version>3.7.0</version>
```
Comment (claussni) [inline pom.xml]: Why did you update this plugin? What was the reason?

File: src/main/java/de/qucosa/dissemination/epicur/EpicurDissMapper.java [hunk 15]
```diff
@@ -0,0 +1,104 @@
+package de.qucosa.dissemination.epicur;
```
Comment (claussni) [inline src/main/java/de/qucosa/dissemination/epicur/EpicurDissMapper.java]: Add missing copyright header. Use the IDEs copyright plugin.

File: src/main/java/de/qucosa/dissemination/epicur/MetadataElementMissing.java [hunk 16]
```diff
@@ -1,4 +1,4 @@
-package de.qucosa.dissemination.epicur.model;
```
Comment (claussni) [inline src/main/java/de/qucosa/dissemination/epicur/MetadataElementMissing.java]: Add copyright header here as well.

File: src/main/java/de/qucosa/dissemination/epicur/UpdateStatus.java [hunk 17]
```diff
@@ -0,0 +1,5 @@
+package de.qucosa.dissemination.epicur;
```
Comment (claussni) [inline src/main/java/de/qucosa/dissemination/epicur/UpdateStatus.java]: Add copyright header.

File: src/test/java/de/qucosa/dissemination/epicur/mapper/EpicurMapperTest.java [hunk 18]
```diff
@@ -0,0 +1,43 @@
+package de.qucosa.dissemination.epicur.mapper;
```
Comment (claussni) [inline src/test/java/de/qucosa/dissemination/epicur/mapper/EpicurMapperTest.java]: Add copyright header.

File: src/test/java/de/qucosa/dissemination/epicur/mapper/EpicurMapperTest.java [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+import org.w3c.dom.Document;
+import org.xml.sax.SAXException;
+
+import de.qucosa.dissemination.epicur.EpicurBuilderException;
+import de.qucosa.dissemination.epicur.EpicurDissMapper;
+
+public class EpicurMapperTest {
+    
+    @Test
+    public void Find_epicur_tag_in_mapped_document() throws ParserConfigurationException, SAXException, IOException, JDOMException, JAXBException, EpicurBuilderException {
```
Comment (claussni) [inline src/test/java/de/qucosa/dissemination/epicur/mapper/EpicurMapperTest.java]: Too many Exceptions in the throws clause.

File: src/test/java/de/qucosa/dissemination/epicur/mapper/EpicurMapperTest.java [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+import java.io.StringWriter;
+
+import javax.xml.bind.JAXBException;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+
+import org.apache.xml.serialize.OutputFormat;
+import org.apache.xml.serialize.XMLSerializer;
+import org.jdom2.JDOMException;
```
Comment (claussni) [inline src/test/java/de/qucosa/dissemination/epicur/mapper/EpicurMapperTest.java]: Where is JDOM2 used? It looks like JDKs XML APIs are sufficient here.

File: .gitignore [hunk 21]
```diff
@@ -1 +1 @@
-/target/
```
Comment (claussni) [inline .gitignore]: What's that change for?

File: src/main/java/de/qucosa/dissemination/epicur/EpicurDissMapper.java [hunk 22]
```diff
@@ -48,13 +70,15 @@ public EpicurDissMapper(String transferUrlPattern, String frontpageUrlPattern, S
...
-            ParserConfigurationException, SAXException, IOException {
+    public Document transformEpicurDiss(Document metsDoc) {
 
         try {
-            this.metsDoc = new SAXBuilder().build(jdom2Build(metsDoc));
-        } catch (JDOMException e) {
+            SAXBuilder saxBuilder = new SAXBuilder();
+            this.metsDoc = saxBuilder.build(jdom2Build(metsDoc));
+        } catch (JDOMException | TransformerException e) {
             logger.error("Cannot JDOM 2 document build.", e);
```
Comment (claussni) [inline src/main/java/de/qucosa/dissemination/epicur/EpicurDissMapper.java]: Change message to "Cannot build JDOM2 document".

File: src/main/java/de/qucosa/dissemination/epicur/EpicurDissMapper.java [hunk 23]
```diff
@@ -48,13 +70,15 @@ public EpicurDissMapper(String transferUrlPattern, String frontpageUrlPattern, S
...
-    public Document transformEpicurDiss(Document metsDoc) throws JAXBException, EpicurBuilderException,
-            ParserConfigurationException, SAXException, IOException {
+    public Document transformEpicurDiss(Document metsDoc) {
 
         try {
-            this.metsDoc = new SAXBuilder().build(jdom2Build(metsDoc));
-        } catch (JDOMException e) {
+            SAXBuilder saxBuilder = new SAXBuilder();
+            this.metsDoc = saxBuilder.build(jdom2Build(metsDoc));
+        } catch (JDOMException | TransformerException e) {
```
Comment (claussni) [inline src/main/java/de/qucosa/dissemination/epicur/EpicurDissMapper.java]: What should happen if the Exceptions get caught? Currently its just logging errors and continues. Throws new exceptions if the method cannot return a Document object.

File: src/main/java/de/qucosa/dissemination/epicur/EpicurDissMapper.java [hunk 24]
```diff
@@ -67,24 +91,35 @@ public Document transformEpicurDiss(Document metsDoc) throws JAXBException, Epic
...
-        DocumentBuilder db = dbf.newDocumentBuilder();
+        DocumentBuilder db;
+        Document epicurDoc = null;
 
-        Document epicurDoc;
-        StringWriter stringWriter = new StringWriter();
-        Marshaller marshaller = JAXBContext.newInstance(Epicur.class).createMarshaller();
-        marshaller.setProperty(Marshaller.JAXB_SCHEMA_LOCATION, XEPICUR_SCHEMA_LOCATION);
-        marshaller.marshal(epicurBuilder.build(), stringWriter);
+        try {
```
Comment (claussni) [inline src/main/java/de/qucosa/dissemination/epicur/EpicurDissMapper.java]: What should happen when Exceptions are caught? Currently its just logging errors and continues. Throws new exceptions if the method cannot return a Document object.

File: src/test/java/de/qucosa/dissemination/epicur/mapper/EpicurMapperTest.java [hunk 25]
```diff
@@ -1,43 +1,58 @@
...
-        EpicurDissMapper mapper = new EpicurDissMapper("http://test.##AGENT##.qucosa.de/id/##PID##", "", "", true);
-        Document epicurRes = mapper.transformEpicurDiss(mets);
-        
-        Assert.assertEquals("epicur", epicurRes.getDocumentElement().getTagName());
-    }
+        DocumentBuilder db;
+        Document mets;
+        Document epicurRes = null;
+
+        try {
```
Comment (claussni) [inline src/test/java/de/qucosa/dissemination/epicur/mapper/EpicurMapperTest.java]: In tests try-catch and log is not needed. In contrast to production code it is better to immediately throw exceptions. 
Add `throws Exception` to the method header instead of listing individual exceptions.
